### PR TITLE
Warn when casting an enum that is fieldless but not C-like

### PR DESCRIPTION
**Pull request by nbdd0121**
Implementation of the second option as mentioned in https://github.com/rust-lang/rust/issues/60553#issuecomment-985007785.

Marked as draft as this needs a lang team decision.

r? @ghost
@rustbot label: +T-lang

### CODE REVIEW

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 1]
```diff
@@ -3633,3 +3634,43 @@ declare_lint! {
...
+    /// Historically we permit casting of enums if none of the variants have
+    /// fields. The intended behaviour is that only C-like (all variants are
+    /// unit variants) enums can be casted this way.
+    ///
+    /// This is a [future-incompatible] lint to transition this to a hard error
+    /// in the future. See [issue #88621] for more details.
+    ///
+    /// [future-incompatible]: ../index.md#future-incompatible-lints
+    /// [issue #88621]: https://github.com/rust-lang/rust/issues/88621
+    pub NOT_CENUM_CAST,
```
Comment (camelid) [inline compiler/rustc_lint_defs/src/builtin.rs:3670]: nit: `CENUM` looks a little off to me. Maybe `C_ENUM` would be better?

Comment (nbdd0121) [reply compiler/rustc_lint_defs/src/builtin.rs:3670]: There is an existing lint "CENUM_IMPL_DROP_CAST" and I just followed that.

**Comment by apiraino** [comment]
I'm tentatively cc'ing `T-lang` to understand if there is an update here. Thanks!

cc @rust-lang/lang

**Comment by nikomatsakis** [comment]
Discussed in the lang team meeting today. Based on the age of this PR, and the fact that we've had several design meetings on this but never come to any clear conclusions, we're going to close this PR for now. We're still blocked on a proposal for a consistent set of rules, essentially. We agree it would be nice to see this resolved! (cc @jswrenn)

**Pull request closed without merging**